### ISSUE TITLE
AvoidUninstantiatedInternalClasses should not be reported for Program classes with async Main

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn-analyzers#1688]
#### Analyzer package

Microsoft.CodeQuality.Analyzers, version [2.6.0](https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers/2.6.0)

#### Analyzer

AvoidUninstantiatedInternalClasses 

#### Repro steps

1. `dotnet new console`
2. `dotnet add package Microsoft.CodeQuality.Analyzers`
3. Add `<LangVersion>7.1</LangVersion>` to the csproj.
3. Change Program.cs to:

    ```c#
    using System.Threading.Tasks;

    static class Program
    {
        static async Task Main()
        {
            await Task.Delay(1);
        }
    }
    ```
4. `dotnet build`

#### Expected behavior

No warning. The `Program` class contains the entry point, so it should not be considered unused.

#### Actual behavior

```
> dotnet build
Microsoft (R) Build Engine version 15.5.180.51428 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  Restore completed in 62,24 ms for C:\code\tmp\hwapp\hwapp.csproj.
Program.cs(7,15): warning CA2007: Do not directly await a Task without calling ConfigureAwait [C:\code\tmp\hwapp\hwapp.csproj]
Program.cs(3,14): warning CA1812: Program is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic). [C:\code\tmp\hwapp\hwapp.csproj]
  hwapp -> C:\code\tmp\hwapp\bin\Debug\netcoreapp2.0\hwapp.dll

Build succeeded.

Program.cs(7,15): warning CA2007: Do not directly await a Task without calling ConfigureAwait [C:\code\tmp\hwapp\hwapp.csproj]
Program.cs(3,14): warning CA1812: Program is an internal class that is apparently never instantiated. If so, remove the code from the assembly. If this class is intended to contain only static members, make it static (Shared in Visual Basic). [C:\code\tmp\hwapp\hwapp.csproj]
    2 Warning(s)
    0 Error(s)

Time Elapsed 00:00:09.96
```

I think the CA2007 above is also invalid; I reported that in a separate issue: https://github.com/dotnet/roslyn-analyzers/issues/1687.
